### ISSUE TITLE
[Bug]: enable_reduce_sample returns incorrect vocabulary IDs for greedy requests in mixed batches

### ISSUE THREAD

**Issue by KeyanHu-git** [vllm-project/vllm-ascend#16396]
### Your current environment

The issue was reproduced with vLLM-Ascend built from source at `fb00c6b`, together with vLLM at the CI-pinned commit `b2f6858`. Key output from `python collect_env.py` in the same test environment:

```text
OS: Ubuntu 22.04.5 LTS (aarch64)
Python: 3.12.13
PyTorch: 2.10.0+cpu
torch-npu: 2.10.0.post4
CANN: 9.1.0
npu-smi: 25.5.2
NPU: Ascend910; 1 and 2 NPUs used to test TP1 and TP2, respectively
```

### 🐛 Describe the bug

With `enable_reduce_sample` enabled during speculative decoding, a batch containing both greedy-sampling and random-sampling requests can cause the greedy requests to return positions in the reordered candidate tensor instead of the corresponding vocabulary token IDs. This produces incorrect output tokens and also affects draft-token acceptance and rejection.

A component-level reproducer uses two requests with entirely finite target logits. Each row is:

```python
[-8., -3., -2., -1., 0., 9., 1., 2.]
```

The maximum corresponds to vocabulary token ID **5**. The first request has temperature `0`; the second has temperature `0` or `0.7`. Each request contains one draft token, with no Top-k/Top-p filtering. The bonus token is fixed to ID `7` so that we can observe whether it is appended; `-1` denotes an output placeholder after rejection.

Results for the first request at TP1:

| Batch temperatures | Draft token ID | Expected output | Actual output |
|---|---:|---|---|
| `[0, 0]` | 0 | `[5, -1]` | `[5, -1]` |
| `[0, 0.7]` | 0 | `[5, -1]` | `[0, 7]` |
| `[0, 0]` | 5 | `[5, 7]` | `[5, 7]` |
| `[0, 0.7]` | 5 | `[5, 7]` | `[0, -1]` |

The same greedy request incorrectly accepts draft token ID `0` and incorrectly rejects draft token ID `5` solely because it is batched with a random-sampling request.

The reproducer and instructions are packaged in `native-reproducer.zip` . The reproducer uses synthetic logits and requires neither model weights nor a service endpoint. Run it in the source environment above without applying the fix:

```bash
TORCH_DEVICE_BACKEND_AUTOLOAD=0 python -m torch.distributed.run \
  --standalone --nnodes=1 --nproc_per_node=1 native_sampling_probe.py \
  --variant before --output-dir results/before-tp1
```

Of the 16 TP1 test cases, all 8 pure-greedy controls pass and all 8 mixed-batch cases fail. Changing `--nproc_per_node` to `2` and using a different output directory reproduces the same type of error at TP2. The JSON reports contain the expected and actual IDs for each case. The script fixes the bonus token and random inputs while executing the actual NPU rejection-sampling kernels and HCCL.

The relevant code is in `vllm_ascend/sample/rejection_sampler.py`. An all-greedy batch skips candidate reordering, whereas a mixed batch passes through `apply_sampling_constraints`, which returns reordered scores and vocabulary indices. The subsequent greedy branch calls `greedy_sample(target_logits)` and interprets column positions using the original vocabulary sharding, without applying the candidate-to-vocabulary mapping.

Related design discussion: #9248. A local fix and regression tests are available; I plan to submit them in a separate PR.

[native-reproducer.zip](https://github.com/user-attachments/files/32138241/native-reproducer.zip)

**Comment by kunpengW-code** [comment]
Thanks for the feedback. The `reduce sampling` feature is experimental and will be deprecated in future releases. We will follow the upstream vLLM community for the `batch-sharded sampling` feature. @SOMEONEUNSEEN Please keep an eye on this issue and its corresponding PR.

**Comment by SOMEONEUNSEEN** [comment]
@zouzy5137 plz take a look

**Comment by zouzy5137** [comment]
Reduce sample will be removed soon. Here is pr #15011 and RFC https://github.com/vllm-project/vllm-ascend/issues/15119
